### PR TITLE
Created Protect Route 

### PR DESCRIPTION
**Pull request by AbeerKaram**
#7
Created Protected Route for user logged and admin only

### CODE REVIEW

File: client/src/context.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  render() {
+    const { isAuth, logged, admin } = this.state;
+    return (
+      <Context.Provider
+        value={{
+          isAuth,
+          logged,
+          admin,
+          setAdmin: (Admin) => this.setState({ admin: Admin }),
+          setLoged: (Loged) => this.setState({ logged: Loged }),
```
Comment (Amoodaa) [inline client/src/context.js:21]: Typo

File: client/src/context.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        }}
+      />
```
Comment (Amoodaa) [inline client/src/context.js:23]: You need to pass children, or else this context cant be passed down
```
      <Context.Provider
        value={{
          isAuth,
          logged,
          admin,
          setAdmin: (Admin) => this.setState({ admin: Admin }),
          setLoged: (Loged) => this.setState({ logged: Loged }),
        }}
      >
        {children}
      </Context.Provider>
```

File: client/src/context.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+
+const Context = React.createContext();
+
+class AuthProvider extends React.Component {
```
Comment (Amoodaa) [inline client/src/context.js:5]: you can 
```suggestion
export class AuthProvider extends React.Component {
```

File: client/src/context.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+          admin,
+          setAdmin: (Admin) => this.setState({ admin: Admin }),
+          setLoged: (Loged) => this.setState({ logged: Loged }),
+        }}
+      />
+    );
+  }
+}
+
+const AuthConsumer = Context.Consumer;
```
Comment (Amoodaa) [inline client/src/context.js:28]: you can 
```suggestion
export const AuthConsumer = Context.Consumer;
```

File: client/src/context.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+const Context = React.createContext();
+
+class AuthProvider extends React.Component {
+  state = {
+    isAuth: false,
+    logged: false,
+    admin: false,
+  };
+
```
Comment (Amoodaa) [inline client/src/context.js:11]: You need to go fetch the user current state: 
 ```
{
   logged: false,
   admin: false
  }
```

and there isnt currently a route, so you have to make one that sends the above data, something like:
`res.send({...req.user})`

File: client/src/context.js (hunk 2 above)
Comment (Amoodaa) [reply client/src/context.js:23]: and you will have to search for the best way to prop-types children

**Comment by Amoodaa** [comment]
One last thing to do as well, is to move the App.js file contents into App/index.js

File: client/src/context.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+
+const Context = React.createContext();
+
+export class AuthProvider extends React.Component {
+  state = {
+    isAuth: false,
```
Comment (Amoodaa) [inline client/src/context.js:7]: What does this `isAuth` mean

File: client/src/App/ProtectedRoute.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import { Redirect, Route } from 'react-router-dom';
+import { AuthConsumer } from '../context';
+
+function ProtectedRoute({ children }) {
+  return (props) => (
```
Comment (Amoodaa) [inline client/src/App/ProtectedRoute.js:7]: This is a function that returns another functions => not a React component

File: client/src/context.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  render() {
+    const { logged, admin } = this.state;
+    return (
+      <Context.Provider
+        value={{
+          logged,
+          admin,
+        }}
+      >
+        {React.Children}
```
Comment (omar-dahdouh) [inline client/src/context.js:34]:
```js
{ this.props.children }
```
you might need to use destructuring
and props validation
```js
  render() {
    const { isAuth, logged, admin } = this.state;
    const { children } = this.props;
    return (
      <Context.Provider
        value={{
          isAuth,
          logged,
          admin,
        }}
      >
        {children}
      </Context.Provider>
    );
  }
}
AuthProvider.propTypes = { children: PropTypes.node.isRequired };
```

Comment (omar-dahdouh) [reply client/src/context.js:34]: also you might need to export in a different way now
```js
export { AuthProvider, AuthConsumer };
```

Comment (Amoodaa) [reply client/src/context.js:34]: she can just
```js
export AuthProvider
```
named exports

File: client/src/context.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+      });
+  }
+
+  render() {
+    const { logged, admin } = this.state;
+    return (
+      <Context.Provider
+        value={{
+          logged,
+          admin,
```
Comment (Amoodaa) [inline client/src/context.js:31]: you will probably need something to be able to edit on this context state, `setAuth`, takes object like:
```js
({ logged, admin }) => setState({ logged, admin })
```

File: client/src/context.js [hunk 10]
```diff
@@ -24,16 +25,21 @@ export class AuthProvider extends React.Component {
...
   render() {
     const { logged, admin } = this.state;
+    const { children } = this.props;
     return (
       <Context.Provider
         value={{
           logged,
           admin,
+          setAdmin: (Admin) => this.setState({ admin: Admin }),
+          setLoged: (Loged) => this.setState({ logged: Loged }),
```
Comment (Amoodaa) [inline client/src/context.js:35]:
```suggestion
          setAuth: ({ logged, admin }) => setState({ logged, admin })
```
make it something like this, it looks better and may help us later on

Comment (Amoodaa) [reply client/src/context.js:35]: for example u need to call setAdmin and then setLoged
unneeded redundancy

**Pull request merged**